src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,322 @@
+                var node = context.LeftToken.Parent;
+
+                var namespacesInScope = GetNamespacesInScope(node, context.SemanticModel, cancellationToken)
+                    .Select(symbol => symbol.ToDisplayString(SymbolDisplayFormats.NameFormat))
+                    .ToImmutableHashSet();
+
+                if (project.SupportsCompilation)
+                {
+                    var builder = ArrayBuilder<TypeImportCompletionItem>.GetInstance();
+
+                    using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+                    {
+                        var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                        var declarationsInCurrentProject = await _typeImportCompletionService
+                                                            .GetAccessibleTopLevelTypesFromProjectAsync(project, namespacesInScope, cancellationToken)
+                                                            .ConfigureAwait(false);
+                        builder.AddRange(declarationsInCurrentProject);
+
+                        foreach (var reference in compilation.References)
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+
+                            if (reference is CompilationReference compilationReference)
+                            {
+                                var declarationsInReference =
+                                    await _typeImportCompletionService.GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                                        project.Solution,
+                                        compilation,
+                                        compilationReference,
+                                        namespacesInScope,
+                                        cancellationToken).ConfigureAwait(false);
+
+                                builder.AddRange(declarationsInReference);
+                            }
+                            else if (reference is PortableExecutableReference peReference)
+                            {
+                                var declarationsInReference = _typeImportCompletionService.GetAccessibleTopLevelTypesFromPEReference(project.Solution,
+                                                                                                                                     compilation,
+                                                                                                                                     peReference,
+                                                                                                                                     namespacesInScope,
+                                                                                                                                     cancellationToken);
```

Comment:
can't read this.

Reply:
both these calls will allocate just to add to a builder.  why not pass the builder in?

Reply:
They only allocate after a cache miss.